exercises/PQ/Practice #1 Opt1/intSimpleList.go

```diff
@@ -0,0 +1,59 @@
+package main
+
+import "errors"
+
+type intSimpleList struct {
+	value int
+	next *intSimpleList
+}
+
+func insert(p *intSimpleList, value int) *intSimpleList {
+	newNode := &intSimpleList{
+		value: value,
+		next: p,
+	}
+
+	return newNode
+}
+
+func getFirst(p *intSimpleList) (int, error) {
+	if p == nil {
+		return 0, errors.New("getFirst failed due to empty list")
+	}
+
+	return p.value, nil
+}
+
+func deleteFirst(p *intSimpleList) (*intSimpleList, error) {
+	if p == nil {
+		return nil, errors.New("deleteFirst failed due to empty list")
+	}
+
+	return p.next, nil
+}
+
+func isMember(p *intSimpleList, value int) bool {
+	if p == nil {
+		return false
+	} else if p.value == value {
+		return true
+	} else {
+		return isMember(p.next, value)
+	}
+}
+
+func copyList(p *intSimpleList) *intSimpleList {
+	if p == nil {
+		return nil
+	} else {
+		return insert(copyList(p.next), p.value)
+	}
+}
```

Comment:
Maybe I'm worng but it seems that if you use the insert function, the copied list is going to have the items in the revers order.
Can you check it?

Reply:
1) I do not know why we are writting in English.
2) The insert statement adds at the beginnig of the list, so the function insert the first element of the source at the begining of the copy of the tail. You can run the main and verify.